src/routes/session.ts

```diff
@@ -4,26 +4,69 @@ import {
 	getUser,
 	comparePassword,
 } from '../platform/auth';
-import { sbvrUtils } from '../platform';
+import { resinApi, root, sbvrUtils } from '../platform';
 import { captureException, handleHttpErrors } from '../platform/errors';
 import { resetCounter } from '../lib/rate-limiting';
 import { RequestHandler } from 'express';
 import { SetupOptions } from '..';
 
 const { BadRequestError, NotFoundError } = sbvrUtils;
 
-export const whoami: RequestHandler = (req, res) =>
-	getUser(req)
-		.then(user =>
-			res.send({ id: user.id, username: user.username, email: user.email }),
-		)
-		.catch(err => {
-			if (handleHttpErrors(req, res, err)) {
-				return;
+export const whoami: RequestHandler = async (req, res) => {
+	try {
+		let user: AnyObject = await getUser(req);
+
+		// If the user email is available, then the record was successfully
+		// populated by the passport middleware
+		if (user.email == null) {
+			// If we get to this point, then the request is most likely being made
+			// using a scoped access token.
+			// First we check to see if this token can access the user resource,
+			// then if it can, we retrieve user document that corresponds to the
+			// `actor` value on the token using root privileges.
+			const count = await resinApi.get({
+				resource: 'user/$count',
```

Comment:
We could get rid fo the `$count` here and add a `$select: 'id'` into the options. So that we already have the id of the potential user.

Reply:
@afitzek this won't work if the scoped access token lets you see multiple users

Reply:
It will work, because in the request we would keep the filter against the actor. But we can save the filter against the actor in the second request and use the primary key. There cannot be two users with the same actor.

Reply:
aha I understand, I'll make the optimization :+1: